erts/emulator/beam/erl_db_util.c

```diff
@@ -2338,8 +2348,13 @@ Eterm db_prog_match(Process *c_p,
             *esp++ = t;
             break;
 	case matchCall0:
-	    bif = (Eterm (*)(Process*, ...)) *pc++;
-	    t = (*bif)(build_proc, bif_args);
+#if defined(__APPLE__) && defined(__aarch64__)
+        bif_call_0 = (BIF_RETTYPE (*)(BIF_ALIST_0)) *pc++;
+        t = (*bif_call_0)(build_proc, esp-1, bif_args);
+#else
+        bif = (Eterm (*)(Process*, ...)) *pc++;
+        t = (*bif)(build_proc, bif_args);
+#endif
```

Comment:
This and the other similar changes should be unconditional.  It is _NEVER_ correct to call a C function via a function pointer type that differs from its original declaration (you can use such casts for transporting the function pointer, but any call must see the correct type).  Abusing this tends to work in older ABIs because pre-C89 K&R C didn't have full prototypes, but newer ABIs are free to handle variadic functions differently from non-variadic ones, as you've dicovered.

Many thanks for working on this.

Reply:
Howdy!

Thank you very much for the suggestion and information! Also a big thanks for y'all getting the PR this was dependent on back ported and merged so damn quickly! 🤘 As I was making the original change, I couldn't figure out why it was done that way previously; so, I'm glad to know it was mostly just abusing older ABIs and not a deeper rabbithole. 

The changes have been made and are now unconditional. I've given it ago on my x86 and ARM64 systems and it works great. Hopefully the change as it is now, is what you had in mind. Please let me know if there's anything else I can do! 

Lastly, I'm still waiting for a good test run on ARM64, not because of BEAM, but because the system will randomly restart (there haven't been any segfaults in the system crash logs though 🎉 ).